### PR TITLE
PDPDEVTOOL-5016 - Unit testing stubs 2

### PR DESCRIPTION
**Pull request by vic-nik**

### CODE REVIEW

File: packages/unit-testing/stubs/ui/dialog/dialog.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+     *
+     */
+    dialog.protptype.create = function (options) { };
+
+    /**
+     * @typedef {Object} ButtonOption
+     * @property {*} label The label of the button
+     * @property {*} value The return value of the dialog if the button was clicked
+     */
+    function ButtonOptionTypedef() {
```
Comment (davidecorreu) [inline packages/unit-testing/stubs/ui/dialog/dialog.js:67]: Not in docs and not exposed but we can leave it I guess.

File: packages/unit-testing/stubs/ui/serverWidget/AssistantStep.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+         * @restriction Server SuiteScript only
+         * @governance none
+         * @param {Object} options
+         * @param {string} options.group  The internal id of the sublist
+         * @return {Array<string>}
+         * @throws {SuiteScriptError} SSS_MISSING_REQD_ARGUMENT when group parameter is missing
+         *
+         * @since 2015.2
+         */
+        this.getSublistFieldIds = function (options) { };
```
Comment (davidecorreu) [inline packages/unit-testing/stubs/ui/serverWidget/AssistantStep.js:59]: Not in table view of docs, but present in tree view. Let's keep it.

File: packages/unit-testing/stubs/ui/serverWidget/Field.js [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+         * @governance none
+         * @param {Object} options
+         * @param {string} options.help The help text for the field
+         * @param {boolean} [options.showInlineForAssistant] This means that field help will appear only in a field help popup box when the field label is clicked. The default value is false, which means the field help appears in a popup when the field label is clicked and does not appear inline.
+         * @return {void}
+         * @throws {SuiteScriptError} SSS_MISSING_REQD_ARGUMENT when help parameter is missing
+         * @throws {SuiteScriptError} SSS_INVALID_TYPE_ARG when help parameter is not string
+         * @since 2015.2
+         */
+        this.setHelpText = function (options) { };
```
Comment (davidecorreu) [inline packages/unit-testing/stubs/ui/serverWidget/Field.js:194]: By the docs seems that this function should return itself, so maybe the return type should be {Field}. Not sure about it.

File: packages/unit-testing/stubs/ui/serverWidget/Form.js [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+        this.title = undefined;
+        /**
+         * This method is called during a beforeLoad UE or a suitelet and the message is later displayed on the client side,
+         * once the pageInit script is completed. The method takes either an already created Message object or the options
+         * object that would be used for creating the message.
+         * User event context can be used to control whether the message is shown on records in view, create, or edit mode (not applicable for Suitelets).
+         * @restriction Server SuiteScript only
+         * @governance none
+         * @param {Object|Message} options config object OR the message options object to be displayed in browser, described in N/ui/message: create()
+         * @param {Message} options.message the message object to be displayed in browser, described in N/ui/message: create()
```
Comment (davidecorreu) [inline packages/unit-testing/stubs/ui/serverWidget/Form.js:29]: This Message is refers to ui/message/MessageInstance, we could inject it.

File: packages/unit-testing/stubs/ui/serverWidget/List.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
+define(['./ListColumn', './Button'], function (ListColumn, Button) {
```
Comment (davidecorreu) [inline packages/unit-testing/stubs/ui/serverWidget/List.js:1]: Missing to inject List.

File: packages/unit-testing/stubs/ui/ui.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+    ui.prototype.dialog = dialog;
+    ui.prototype.message = message;
+    ui.prototype.serverWidget = serverWidget;
+
+    /**
+     * @exports N/ui
+     * @namespace ui
+     */
+    return new ui();
```
Comment (davidecorreu) [inline packages/unit-testing/stubs/ui/ui.js:19]: Adding this `ui` moudule makes me think if we should also add the `N` module at the root of our stubs. It's just an open question to consider later with the team.

File: packages/unit-testing/stubs/ui/dialog/dialog.js (hunk 1 above)
Comment (vic-nik) [reply packages/unit-testing/stubs/ui/dialog/dialog.js:67]: I will remove it, looks like an internal definition

File: packages/unit-testing/stubs/ui/serverWidget/Field.js (hunk 3 above)
Comment (vic-nik) [reply packages/unit-testing/stubs/ui/serverWidget/Field.js:194]: Not sure either, could be inconsistency in the doc as well. Looks like the functions that return Field object are mostly related to the placement of the field in the UI and this one is different. I am inclined to leave it as is for now.

File: packages/unit-testing/stubs/ui/serverWidget/Form.js (hunk 4 above)
Comment (vic-nik) [reply packages/unit-testing/stubs/ui/serverWidget/Form.js:29]: Yes

File: packages/unit-testing/stubs/ui/ui.js (hunk 6 above)
Comment (vic-nik) [reply packages/unit-testing/stubs/ui/ui.js:19]: It made sense to include N/ui becuase of this Help Center section:
https://nlcorp.app.netsuite.com/app/help/helpcenter.nl?fid=section_156939471908.html
Which says:
Use [‘N/ui’] as the first argument of the define function as a shortcut to load the three modules (N/ui/dialog, N/ui/message, and N/ui/serverWidget) on server scripts, or the N/ui/dialog and N/ui/message on client scripts.

I would avoid creating N module as its usage is not recommended in SuiteScript. However, if we find similar examples in the doc for other sub-modules, we should expose them. E.g. N/crypto and N/crypto/certificate

**Review by alii-gator — APPROVED** [review]

**Pull request merged**